# Онлайн библиотека, SQL проект

## Описание данных и задание SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.



Таблица `books` содержит данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.


Таблица `authors` cодержит данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.


Таблица `publishers` cодержит данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

Таблица `ratings` cодержит данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица `reviews` cодержит данные о пользовательских обзорах на книги:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Как выполнить задание?
* Опишите цели исследования;
* Исследуйте таблицы — выведите первые строки, посчитайте количество стро в каждой таблице;
* Сделайте по одному SQL-запросу для решения каждого задания;
* Выведите результат каждого запроса в тетрадке;
* Опишите выводы по каждой из решённых задач.

## Цели исследования: 

Проанализировать базу данных онлайн библиотеки, ознакомиться с данными и выполнить sql-запросы для формирования целостного представления о сервисе. Сформировать  ценностное предложение для нового продукта.


In [1]:
# импортируем библиотеки
import pandas as pd
import sqlalchemy as sa

In [2]:
# устанавливаем параметры
db_config = {
            'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db' # название базы данных
            }
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    '''Открываем соединение, получаем данные из sql, закрываем соединение'''
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)

## Задания

1) Посчитайте, сколько книг вышло после 1 января 2000 года;
2) Для каждой книги посчитайте количество обзоров и среднюю оценку;
3) Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4) Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5) Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.


## Знакомство с данными

In [4]:
# первые 5 строчк таблицы boooks
query = '''
            SELECT * 
            FROM books 
            LIMIT 5
        '''
get_sql_data(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
# первые 5 строчк таблицы authors
query = '''  
            SELECT * 
            FROM authors 
            LIMIT 5
        '''
get_sql_data(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
# первые 5 строчк таблицы ratings
query = ''' 
            SELECT * 
            FROM ratings 
            LIMIT 5
        '''
get_sql_data(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
# первые 5 строчк таблицы reviews
query = '''
            SELECT * 
            FROM reviews 
            LIMIT 5
        '''
get_sql_data(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [8]:
# первые 5 строчк таблицы publishers
query = ''' 
            SELECT * 
            FROM publishers 
            LIMIT 5
        '''
get_sql_data(query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

## Выполнение заданий

### Задание 1. Посчитайте, сколько книг вышло после 1 января 2000 года;

In [9]:
# первые 5 строчк таблицы boooks
query = '''
                SELECT COUNT(book_id) 
                FROM books 
                WHERE publication_date > '2000-01-01'
        '''
get_sql_data(query)

count
0    819

### Вывод 

После 1 января 2000 года, согласно данным из базы данных, вышло `819` книг

### Задание 2. Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [10]:
query = '''
              SELECT b.book_id,
                     b.title,
                     ROUND (AVG(rtng.rating), 2) AS avg_rating,
                     COUNT(DISTINCT rvws.review_id) AS review_count
              FROM books AS b
              LEFT JOIN reviews AS rvws ON b.book_id = rvws.book_id
              LEFT JOIN ratings AS rtng ON b.book_id = rtng.book_id
              GROUP BY b.book_id
              ORDER BY review_count DESC
        '''
get_sql_data(query)

book_id                                              title  avg_rating  \
0        948                            Twilight (Twilight  #1)        3.66   
1        963                                Water for Elephants        3.98   
2        734                                   The Glass Castle        4.21   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...        4.41   
4        695  The Curious Incident of the Dog in the Night-Time        4.08   
..       ...                                                ...         ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel        3.67   
996      808                            The Natural Way to Draw        3.00   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites        5.00   
998      221                          Essential Tales and Poems        4.00   
999      191  Disney's Beauty and the Beast (A Little Golden...        4.00   

     review_count  
0               7  
1               6  
2               6  
3               6  
4               6  
..            ...  
995             0  
996             0  
997             0  
998             0  
999             0  

[1000 rows x 4 columns]

### Вывод

В базе данных максимально количество отзывов на одну книгу - `7`, книга - `Twilight (Twilight #1)`, так же, в базе данных наблюдаются книги без отзывов

### Задание 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [11]:
query = '''
            
              
            WITH
            b as (
                SELECT publisher_id,
                        book_id,
                        num_pages
                FROM books
                WHERE num_pages > 50),
            p as (
                SELECT *
                FROM publishers
            )
            SELECT COUNT(b.book_id),
                    p.publisher
            FROM b
            LEFT JOIN p ON b.publisher_id = p.publisher_id
            GROUP BY p.publisher
            ORDER BY count DESC
            LIMIT 1
            

        '''
get_sql_data(query)

count      publisher
0     42  Penguin Books

### Вывод

Согласно данным из базы данных, больше всего издательств, толще `50` страниц, было у издательства `Penguin Books`

### Задание 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [12]:
query = '''
        WITH
            r AS (  
                    SELECT book_id,
                    COUNT(rating_id) as cnt_rating,
                    AVG(rating) as avg_rating
                    FROM ratings
                    GROUP BY book_id
                    HAVING  COUNT(rating_id) >= 50),
            b AS (
                    SELECT book_id,
                            author_id
                    FROM books), 
            a AS (
                    SELECT author_id,
                            author
                    FROM authors)

            SELECT ROUND(AVG(r.avg_rating), 2) as total_avg,
                    a.author
            FROM r
            LEFT JOIN b ON r.book_id = b.book_id
            LEFT JOIN a ON b.author_id = a.author_id
            GROUP BY a.author
            ORDER BY total_avg DESC
            LIMIT 1;
                
        '''
get_sql_data(query)

total_avg                      author
0       4.28  J.K. Rowling/Mary GrandPré

### Вывод
Данные из базы данных говорят, что самый высоко оцениваемый автор, среди авторов, книги которых набрали `50` и более оценок является `J.K. Rowling/Mary GrandPré`

### Задание 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [13]:
query = '''
            
        SELECT AVG(f.cnt_review) AS average_reviews
        FROM (
            WITH 
            rtng AS (
                SELECT DISTINCT username, 
                       COUNT(rating) as cnt_rating
                FROM ratings
                GROUP BY username
                HAVING COUNT(rating) > 48
            ),
            rvws AS (
                SELECT DISTINCT username, 
                       COUNT(review_id) as cnt_review
                FROM reviews
                GROUP BY username
            )
            
            SELECT *
            FROM rtng
            LEFT JOIN rvws ON rtng.username = rvws.username) as f

        '''
get_sql_data(query)

average_reviews
0             24.0

### Вывод
В базе данных сервиса хранится информация, согласно которой, активные пользователи, поставившие больше `48` оценок, в среднем делают `24` обзора

## Итоговый вывод

Проанализировав базу данных крупной онлайн библиотеки по чтению книг, мы, как аналитики, можем сделать следующие выводы:

   * в нашей базе данных наблюдается `819` книг, вышедших в 21 веке
    
   * на первый взгляд, пользователи не очень охотно оставляют обзоры, так, очевидно, бестселлер своего времени, `Twilight (Twilight #1)` собрал всего 7 обзоров. А некоторые книги и вовсе не имеют обзоров.
    
   * Самое продуктивное издательство - `Penguin Books`, на его счету выпущено `42` книги, толще `50` страниц
    
   * Самый высоко оцениваемый автор - `J.K. Rowling/Mary GrandPré`, средняя оценка `4.28`
    
   * В приложении есть активные пользователи, которые поставили больше 48 оценок и их среднее количество оставляемых обзоров - `24`
   * Так же стоит отметить, что в данные в базе, на первый взгляд, чистые, названия таблиц и столбцов и их тип данных совпадает с предоставленной er-диаграммой